python/rmm/_lib/memory_resource.pyx

```diff
@@ -444,6 +464,62 @@ cdef class BinningMemoryResource(UpstreamResourceAdaptor):
                     bin_resource.get_mr())
 
 
+cdef void* _allocate_callback_wrapper(
+    size_t nbytes,
+    cuda_stream_view stream,
+    void* ctx
+) with gil:
+    return <void*><uintptr_t>((<object>ctx)(nbytes))
+
+cdef void _deallocate_callback_wrapper(
+    void* ptr,
+    size_t nbytes,
+    cuda_stream_view stream,
+    void* ctx
+) with gil:
+    (<object>ctx)(<uintptr_t>(ptr), nbytes)
+
+
+cdef class CallbackMemoryResource:
+    """
+    A memory resource that uses the user-provided callables to do
+    memory allocation and deallocation.
+
+    The allocation function must accept a single integer argument,
+    representing the number of bytes to allocate, and return an
+    integer representing the pointer to the allocated memory.
+
+    The deallocation function must accept two arguments, an integer
+    representing the pointer to the memory to free, and a second
+    integer representing the number of bytes to free.
+
+    ``CallbackMemoryResource`` should really only be used for
+    debugging memory issues, as there is a significant performance
+    penalty associated with using a Python function for each memory
+    allocation and deallocation.
+
+    Parameters
+    ----------
+    allocate_func: callable
+    deallocate_func: callable
```

Comment:
The lines about the allocation and deallocation functions in the description above should probably go here in the parameter descriptions.